File: sdk/android/tokenpay-id-sdk/src/main/java/space/tokenpay/id/TpidAuth.kt
package space.tokenpay.id

import android.content.Context
import android.util.Log
import space.tokenpay.id.internal.ApiClient
import space.tokenpay.id.internal.SecureStorage
import space.tokenpay.id.internal.SdkConfigLoader
import space.tokenpay.id.internal.Telemetry
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import java.util.concurrent.atomic.AtomicReference

/**
 * Entry point for the TOKEN PAY ID SDK.
 *
 * Call [initialize] once from your [android.app.Application.onCreate], then use
 * [TpidLoginButton] in Compose or [TpidWidget.present] programmatically.
 */
object TpidAuth {

    private const val TAG = "TpidAuth"
    private val stateRef = AtomicReference<State?>(null)
    internal val scope = CoroutineScope(SupervisorJob() + Dispatchers.Default)

    internal data class State(
        val appContext: Context,
        val config: TpidConfig,
        val api: ApiClient,
        val storage: SecureStorage,
        val telemetry: Telemetry,
        val sdkConfigLoader: SdkConfigLoader,
    )

    /**
     * Initialize the SDK. Safe to call multiple times — subsequent calls replace config.
     */
    @JvmStatic
    fun initialize(context: Context, config: TpidConfig) {
        val appCtx = context.applicationContext
        val storage = SecureStorage(appCtx, config.clientId)
        val api = ApiClient(
            issuer = config.issuer,
            pinTls = config.pinTlsCertificates,
            storage = storage,
        )
        val telemetry = Telemetry(
            api = api,
            clientId = config.clientId,
            enabled = config.enableTelemetry,
        )
        val loader = SdkConfigLoader(api = api, storage = storage)
        stateRef.set(State(appCtx, config, api, storage, telemetry, loader))
        Log.i(TAG, "TOKEN PAY ID SDK initialized (v${BuildConfig.TPID_SDK_VERSION}) for client=${config.clientId.take(15)}…")

        // Fire-and-forget: prefetch SDK config (branding, TLS pins) in background.
        scope.launch(Dispatchers.IO) { runCatching { loader.refreshConfig(config.clientId) } }
    }

    internal fun requireState(): State = stateRef.get()
        ?: throw TpidError.ConfigurationError("TpidAuth.initialize() not called. Call it from Application.onCreate().")

    /** Whether the user has a valid (non-expired) stored session. */
    @JvmStatic
    fun hasValidSession(): Boolean {
        val s = stateRef.get() ?: return false
        val tok = s.storage.readTokens() ?: return false
        return tok.expiresAtMs > System.currentTimeMillis() + 30_000
    }

    /** Returns the cached access token if valid, otherwise attempts refresh. Returns null if not signed in. */
    @JvmStatic
    suspend fun getAccessToken(): String? {
        val s = stateRef.get() ?: return null
        val tok = s.storage.readTokens() ?: return null
        if (tok.expiresAtMs > System.currentTimeMillis() + 30_000) return tok.accessToken
        // Try refresh
        val rt = tok.refreshToken ?: return null
        val refreshed = runCatching { s.api.refreshToken(rt, s.config.clientId) }.getOrNull()
            ?: return null
        s.storage.writeTokens(refreshed)
        return refreshed.accessToken
    }

    /**
     * Sign out. Clears local tokens and optionally revokes on the server.
     */
    @JvmStatic
    fun signOut(revokeOnServer: Boolean = true) {
        val s = stateRef.get() ?: return
        val tokens = s.storage.readTokens()
        s.storage.clearTokens()
        if (revokeOnServer && tokens?.refreshToken != null) {
            scope.launch(Dispatchers.IO) {
                runCatching { s.api.revokeToken(tokens.refreshToken, s.config.clientId) }
            }
        }
        s.telemetry.track("widget.sign_out")
    }

    /**
     * Start a Device Flow (RFC 8628). Returns an object containing the user code
     * and the verification URL to display on-screen (e.g. on Android TV).
     */
    @JvmStatic
    suspend fun startDeviceFlow(): DeviceFlowSession {
        val s = requireState()
        val resp = s.api.deviceAuthorize(s.config.clientId, s.config.scopeString)
        return DeviceFlowSession(
            deviceCode = resp.deviceCode,
            userCode = resp.userCode,
            verificationUri = resp.verificationUri,
            verificationUriComplete = resp.verificationUriComplete,
            expiresIn = resp.expiresIn,
            interval = resp.interval,
        )
    }

    /**
     * Poll the Device Flow token endpoint until authorization is approved, denied, or expires.
     * Suspends for up to [DeviceFlowSession.expiresIn] seconds.
     */
    @JvmStatic
    suspend fun pollDeviceFlow(session: DeviceFlowSession): TpidResult {
        val s = requireState()
        return s.api.pollDeviceToken(session, s.config.clientId).also { r ->
            if (r is TpidResult.Success) {
                s.storage.writeTokens(
                    SecureStorage.Tokens(
                        accessToken = r.accessToken,
                        refreshToken = r.refreshToken,
                        idToken = r.idToken,
                        expiresAtMs = System.currentTimeMillis() + r.expiresIn * 1000L,
                        scope = r.scope,
                    )
                )
            }
        }
    }

    /**
     * Return the currently-stored user profile if available (reads the last tokens' stored user).
     */
    @JvmStatic
    fun currentUser(): TpidUser? = stateRef.get()?.storage?.readTokens()?.user
}

data class DeviceFlowSession(
    val deviceCode: String,
    val userCode: String,
    val verificationUri: String,
    val verificationUriComplete: String,
    val expiresIn: Int,
    val interval: Int,
)

File: sdk/swift/Sources/TokenPayID/TpidAuth.swift
import Foundation

/// Entry point for TOKEN PAY ID SDK.
///
/// Call `TpidAuth.shared.initialize(config:)` once at app launch, then use
/// `TpidLoginButton` in SwiftUI or `TpidWidget.present(...)` programmatically.
public final class TpidAuth: @unchecked Sendable {

    public static let shared = TpidAuth()
    private init() {}

    internal struct State {
        let config: TpidConfig
        let api: ApiClient
        let storage: SecureStorage
        let telemetry: Telemetry
        let sdkConfigLoader: SdkConfigLoader
    }

    private let lock = NSLock()
    private var _state: State?

    internal var state: State? {
        lock.lock(); defer { lock.unlock() }
        return _state
    }

    /// Initialize the SDK. Safe to call multiple times; subsequent calls replace config.
    public func initialize(config: TpidConfig) {
        let storage = SecureStorage(clientId: config.clientId)
        let api = ApiClient(issuer: config.issuer, pinTLS: config.pinTLSCertificates, storage: storage)
        let telemetry = Telemetry(api: api, clientId: config.clientId, enabled: config.enableTelemetry)
        let loader = SdkConfigLoader(api: api, storage: storage)

        lock.lock()
        _state = State(config: config, api: api, storage: storage, telemetry: telemetry, sdkConfigLoader: loader)
        lock.unlock()

        Task { try? await loader.refreshConfig(clientId: config.clientId) }
    }

    internal func requireState() throws -> State {
        guard let s = state else {
            throw TpidError.configurationError(reason: "TpidAuth not initialized. Call TpidAuth.shared.initialize(config:) at app launch.")
        }
        return s
    }

    /// True if we have a cached, non-expired session.
    public var hasValidSession: Bool {
        guard let s = state, let session = s.storage.readSession() else { return false }
        return session.isValid
    }

    /// Return a valid access token, refreshing via refresh_token if needed. Returns nil if user is not signed in.
    public func getAccessToken() async throws -> String? {
        guard let s = state, let session = s.storage.readSession() else { return nil }
        if session.isValid { return session.accessToken }
        guard let refresh = session.refreshToken else { return nil }
        do {
            let refreshed = try await s.api.refreshToken(refreshToken: refresh, clientId: s.config.clientId)
            var newSession = refreshed
            if newSession.user.id.isEmpty { newSession = TpidSession(accessToken: refreshed.accessToken, refreshToken: refreshed.refreshToken, idToken: refreshed.idToken, expiresAt: refreshed.expiresAt, tokenType: refreshed.tokenType, scope: refreshed.scope, user: session.user) }
            try? s.storage.writeSession(newSession)
            return newSession.accessToken
        } catch {
            return nil
        }
    }

    /// Sign out, clearing the Keychain and optionally revoking server-side.
    public func signOut(revokeOnServer: Bool = true) {
        guard let s = state else { return }
        let tokens = s.storage.readSession()
        s.storage.clearSession()
        if revokeOnServer, let t = tokens?.refreshToken {
            Task { await s.api.revokeToken(t, clientId: s.config.clientId) }
        }
        s.telemetry.track("widget.sign_out")
    }

    /// Currently-stored user profile (from last successful login).
    public func currentUser() -> TpidUser? {
        state?.storage.readSession()?.user
    }

    // MARK: - Device Flow

    public func startDeviceFlow() async throws -> TpidDeviceFlowSession {
        let s = try requireState()
        let r = try await s.api.deviceAuthorize(clientId: s.config.clientId, scope: s.config.scopeString)
        return TpidDeviceFlowSession(
            deviceCode: r.deviceCode,
            userCode: r.userCode,
            verificationURI: r.verificationURI,
            verificationURIComplete: r.verificationURIComplete,
            expiresIn: r.expiresIn,
            interval: r.interval
        )
    }

    public func pollDeviceFlow(_ session: TpidDeviceFlowSession) async throws -> TpidSession {
        let s = try requireState()
        let result = try await s.api.pollDeviceToken(session: session, clientId: s.config.clientId)
        try? s.storage.writeSession(result)
        return result
    }
}
